### PR TITLE
[ZEPPELIN-3033] IPYNB import/export

### PR DESCRIPTION
**Pull request by ShamsUlAzeem**
### What is this PR for?
Ability to import from and export to notebooks in jupyter ipynb format. This is helpful if you want users to be able to see rendered notebooks in nbviewer on github.

### What type of PR is it?
Feature

### Todos
- Saving 'results' part from json to ipynb
- Using the default interpreter magic for code, according to the default interpreter
- Working on writing tests using travis CI

### What is the Jira issue?
ZEPPELIN-3033

### How should this be tested?
- Working on writing tests using travis CI

### Questions:
* Does the licenses files need update? 
     **No**
* Is there breaking changes for older versions?
     **No**
* Does this needs documentation?
     **No**

### CODE REVIEW

**Comment by jongyoul** [comment]
Looks like a nice feature

**Comment by miandreu** [comment]
It is a important feature, I hope everything goes well

File: zeppelin-interpreter/pom.xml [hunk 1]
```diff
@@ -32,7 +32,7 @@
   <artifactId>zeppelin-interpreter</artifactId>
   <packaging>jar</packaging>
   <version>0.8.0-SNAPSHOT</version>
-  <name>Zeppelin: Interpreter</name>
+  <name>zeppelin-interpreter</name>
```
Comment (felixcheung) [inline zeppelin-interpreter/pom.xml]: why the change? is this related to ZEPPELIN-3025?

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java [hunk 2]
```diff
@@ -167,6 +164,100 @@ private Gson getGson(GsonBuilder gsonBuilder) {
...
+    Note noteFormat = getGson(new GsonBuilder()).fromJson(note, Note.class);
+
+    JsonObject nbformat = new JsonObject();
+    JsonArray cells = new JsonArray();
+
+    RegularExpression MD = new RegularExpression("%md\\s");
+    RegularExpression SQL = new RegularExpression("%sql\\s");
+    RegularExpression UNKNOWN_MAGIC = new RegularExpression("%\\w+\\s");
+    RegularExpression HTML = new RegularExpression("%html\\s");
+    RegularExpression SPARK = new RegularExpression("%spark\\s");
```
Comment (felixcheung) [inline zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: hmm, I think perhaps we need more flexibility here - for example the group/interpreter name for Spark could be changed. Same for all interpreter names like md for example.

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java [hunk 3]
```diff
@@ -167,6 +164,100 @@ private Gson getGson(GsonBuilder gsonBuilder) {
...
+
+      cells.add(codeJson);
+
+      index++;
+    }
+
+    JsonObject metadataJson = new JsonObject();
+
+    JsonObject kernelspecJson = new JsonObject();
+    kernelspecJson.addProperty("display_name", "Spark 2.0.0 - Scala 2.11");
```
Comment (felixcheung) [inline zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: how is this going to be used? I don't think we should hardcode the Spark version number and Scala version

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java [hunk 4]
```diff
@@ -167,6 +164,100 @@ private Gson getGson(GsonBuilder gsonBuilder) {
...
+    kernelspecJson.addProperty("language", "scala");
+    kernelspecJson.addProperty("name", "spark2-scala");
+
+    JsonObject languageInfoJson = new JsonObject();
+    languageInfoJson.addProperty("codemirror_mode", "text/x-scala");
+    languageInfoJson.addProperty("file_extension", ".scala");
+    languageInfoJson.addProperty("mimetype", "text/x-scala");
+    languageInfoJson.addProperty("name", "scala");
+    languageInfoJson.addProperty("pygments_lexer", "scala");
+    languageInfoJson.addProperty("version", "2.11.8");
```
Comment (felixcheung) [inline zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: ditto

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/socket/Message.java [hunk 5]
```diff
@@ -65,6 +65,8 @@
 
     NOTE_RENAME,
 
+    CONVERT_NOTE,     // converting note to nbformat
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/socket/Message.java]: I'd extend this to include nbformat, eg. CONVERT_NOTE_NBFORMAT

File: zeppelin-interpreter/pom.xml (hunk 1 above)
Comment (ShamsUlAzeem) [reply zeppelin-interpreter/pom.xml]: Yup! Looks like it's fixed now in master...

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/socket/Message.java (hunk 5 above)
Comment (ShamsUlAzeem) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/socket/Message.java]: Done

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java [hunk 6]
```diff
@@ -16,28 +16,25 @@
...
 import com.google.gson.typeadapters.RuntimeTypeAdapterFactory;
+import com.sun.org.apache.xerces.internal.impl.xpath.regex.RegularExpression;
 import org.apache.commons.cli.CommandLine;
 import org.apache.commons.cli.CommandLineParser;
 import org.apache.commons.cli.DefaultParser;
 import org.apache.commons.cli.HelpFormatter;
 import org.apache.commons.cli.Options;
 import org.apache.commons.cli.ParseException;
 
+import org.apache.commons.lang3.StringUtils;
```
Comment (mgaido91) [inline zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: I think this should go before the upper newline

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java [hunk 7]
```diff
@@ -167,6 +164,100 @@ private Gson getGson(GsonBuilder gsonBuilder) {
...
+
+    int index = 0;
+    for (Paragraph paragraph : noteFormat.getParagraphs()) {
+      String code = StringUtils.stripStart(paragraph.getText(), " ");
+      JsonObject codeJson = new JsonObject();
+
+      if (code == null || code.trim().isEmpty())
+        continue;
+
+      System.out.println(code);
```
Comment (mgaido91) [inline zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: why?

Comment (ShamsUlAzeem) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: Oops! I forgot to remove this line. Good catch

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 6 above)
Comment (ShamsUlAzeem) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: Yup! I've updated the code

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 2 above)
Comment (ShamsUlAzeem) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: Don't know what you mean here exactly. Do you want me to include conversions for other interpreters like python and sh?

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 3 above)
Comment (ShamsUlAzeem) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: Yeah I commented them as they weren't going to be used: https://github.com/apache/zeppelin/pull/2649/commits/deadcc77475c15b73470a87c9c07b4dd68b9236b

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 4 above)
Comment (ShamsUlAzeem) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: https://github.com/apache/zeppelin/pull/2649/commits/deadcc77475c15b73470a87c9c07b4dd68b9236b

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 2 above)
Comment (felixcheung) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: I mean in Zeppelin I could have multiple "names" for the Spark interpreter, it might not be `%spark`

File: zeppelin-interpreter/pom.xml (hunk 1 above)
Comment (ShamsUlAzeem) [reply zeppelin-interpreter/pom.xml]: I merged the master branch and renamed the project name https://github.com/apache/zeppelin/pull/2649/commits/9fc64c1d2a4a8de136e63c224a8b1c392902e488

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 2 above)
Comment (ShamsUlAzeem) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: How can I access the potentially changed name of the interpreter here?

**Comment by tromika** [comment]
Any update on this?

**Comment by ShamsUlAzeem** [comment]
@tromika I needed more info on how to access the interpreter group name. As soon as I get it I can add the complete the functionality. @felixcheung Any help on that part?

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 2 above)
Comment (tromika) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: would be great to merge this feature so bumping this question :) @ShamsUlAzeem @felixcheung

**Comment by Tagar** [comment]
Would be great to have this functionality part of core Zeppelin.
Any plans to move this PR forward? Thanks!

File: zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java (hunk 2 above)
Comment (felixcheung) [reply zeppelin-jupyter/src/main/java/org/apache/zeppelin/jupyter/JupyterUtil.java]: Sorry about the delay. Each interpreter can have multiple instances and in more than one groups and/or have more than one names. (see config example [here](https://github.com/apache/zeppelin/blob/fe07e5a4962f9457fc2aca473511a50a553499b0/spark/interpreter/src/main/resources/interpreter-setting.json#L171) `group` and `name`fields)

One approach is to:
1. get the binded interpreter instead of parsing the text (https://github.com/apache/zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java#L241)

2. from the interpreter instance, get the classname (https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/Interpreter.java#L240)

3. check the classname against the well-known built-in ones like `org.apache.zeppelin.spark.PySparkInterpreter` (https://github.com/apache/zeppelin/blob/fe07e5a4962f9457fc2aca473511a50a553499b0/spark/interpreter/src/main/resources/interpreter-setting.json#L171)

**Comment by jongyoul** [comment]
@ShamsUlAzeem Do you have any progress on this? I like this feature. :-)

**Comment by ShamsUlAzeem** [comment]
@jongyoul Sorry I missed @felixcheung 's comment. I'll start my work on this PR again

**Comment by jongyoul** [comment]
@ShamsUlAzeem GREAT!! Thank you!!

**Comment by lilian-benoit** [comment]
@ShamsUlAzeem  I like too this feature. Any progress ?

**Comment by zjffdu** [comment]
ping @ShamsUlAzeem Could you resolve the conflict ? I will help to review it after that.

**Comment by ShamsUlAzeem** [comment]
@zjffdu Sure. I'll complete this feature this weekend.

**Comment by ShamsUlAzeem** [comment]
@zjffdu resolve the conflicts. Not entirely sure how the changes would take effect in the newer zeppelin since this PR is quite old but I've resolved them as best as I could. Let me know what I should add more. I'll test the functionality later when I get more time for this.

**Comment by ShamsUlAzeem** [comment]
Ah shit! Some unintended changes got creeped in. Stupid IDE... I'll do an update for this

**Comment by ShamsUlAzeem** [comment]
@zjffdu oh much better now. You can have a look at it. 

There's this comment about interpreters and interpreter groups by @felixcheung that I want to look at later.

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 8]
```diff
@@ -1128,9 +1135,24 @@ public void onSuccess(Note note, ServiceContext context) throws IOException {
...
+            .put("name", fromMessage.get("name"));
+
+    conn.send(serializeMessage(resp));
+  }
+          
   protected Note importNote(NotebookSocket conn, Message fromMessage) throws IOException {
+    String noteJson;
     String noteName = (String) ((Map) fromMessage.get("note")).get("name");
-    String noteJson = gson.toJson(fromMessage.get("note"));
+    if (((Map) fromMessage.get("note")).get("cells") == null) {
```
Comment (zjffdu) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Add one line of comment to explain how we determine whether it is zeppelin note or jupyter note

**Comment by zjffdu** [comment]
@ShamsUlAzeem Please setup your travis to enable the CI. See guideline here. https://zeppelin.apache.org/contribution/contributions.html#continuous-integration

I tried to build your PR, but hit maven dependency  conflicts issue.
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.3.1:enforce (enforce) on project zeppelin-server: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
[ERROR] Failed while enforcing releasability the error(s) are [
[ERROR] Dependency convergence error for org.ow2.asm:asm:5.0.4 paths to dependency are:
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-zengine:0.9.0-SNAPSHOT
[ERROR]     +-org.apache.zeppelin:zeppelin-interpreter:0.9.0-SNAPSHOT
[ERROR]       +-io.atomix:atomix:3.0.0-rc4
[ERROR]         +-io.atomix:atomix-utils:3.0.0-rc4
[ERROR]           +-com.esotericsoftware:kryo:4.0.2
[ERROR]             +-com.esotericsoftware:reflectasm:1.11.3
[ERROR]               +-org.ow2.asm:asm:5.0.4
[ERROR] and
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-jupyter:0.9.0-SNAPSHOT
[ERROR]     +-org.apache.zeppelin:zeppelin-markdown:0.9.0-SNAPSHOT
[ERROR]       +-org.pegdown:pegdown:1.6.0
[ERROR]         +-org.parboiled:parboiled-java:1.1.7
[ERROR]           +-org.ow2.asm:asm:5.0.3
[ERROR] and
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-jupyter:0.9.0-SNAPSHOT
[ERROR]     +-org.apache.zeppelin:zeppelin-markdown:0.9.0-SNAPSHOT
[ERROR]       +-org.pegdown:pegdown:1.6.0
[ERROR]         +-org.parboiled:parboiled-java:1.1.7
[ERROR]           +-org.ow2.asm:asm-tree:5.0.3
[ERROR]             +-org.ow2.asm:asm:5.0.3
[ERROR] and
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-com.nimbusds:nimbus-jose-jwt:4.41.2
[ERROR]     +-net.minidev:json-smart:2.3
[ERROR]       +-net.minidev:accessors-smart:1.2
[ERROR]         +-org.ow2.asm:asm:5.0.4
[ERROR] ,
[ERROR] Dependency convergence error for com.google.guava:guava:22.0 paths to dependency are:
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-zengine:0.9.0-SNAPSHOT
[ERROR]     +-org.apache.zeppelin:zeppelin-interpreter:0.9.0-SNAPSHOT
[ERROR]       +-io.atomix:atomix:3.0.0-rc4
[ERROR]         +-io.atomix:atomix-utils:3.0.0-rc4
[ERROR]           +-com.google.guava:guava:22.0
[ERROR] and
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-jupyter:0.9.0-SNAPSHOT
[ERROR]     +-com.google.guava:guava:15.0
[ERROR] ,
[ERROR] Dependency convergence error for org.apache.commons:commons-lang3:3.7 paths to dependency are:
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-zengine:0.9.0-SNAPSHOT
[ERROR]     +-org.apache.zeppelin:zeppelin-interpreter:0.9.0-SNAPSHOT
[ERROR]       +-org.apache.commons:commons-lang3:3.7
[ERROR] and
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-zengine:0.9.0-SNAPSHOT
[ERROR]     +-org.apache.commons:commons-lang3:3.7
[ERROR] and
[ERROR] +-org.apache.zeppelin:zeppelin-server:0.9.0-SNAPSHOT
[ERROR]   +-org.apache.zeppelin:zeppelin-jupyter:0.9.0-SNAPSHOT
[ERROR]     +-org.apache.zeppelin:zeppelin-markdown:0.9.0-SNAPSHOT
[ERROR]       +-org.apache.commons:commons-lang3:3.4
[ERROR] ]
[ERROR] -> [Help 1]
```

**Comment by ShamsUlAzeem** [comment]
@zjffdu ah alright. on it

**Comment by zjffdu** [comment]
@ShamsUlAzeem Have you configured your travis to enable the CI ? https://zeppelin.apache.org/contribution/contributions.html#continuous-integration

**Comment by ShamsUlAzeem** [comment]
@zjffdu yup I configured it a long time ago and it's working now as well. I'll do a couple of changes and run some tests before notifying you for updates.

**Comment by zjffdu** [comment]
Thanks @ShamsUlAzeem Seems the CI is still running, let's wait for the CI result.

**Comment by ShamsUlAzeem** [comment]
@zjffdu Cool! But I haven't resolved the issues yet. I'll let you know when they're resolved. I just did a random commit to check the CI status

**Comment by ShamsUlAzeem** [comment]
@zjffdu solved the dependency issues and some zeppelin-web lint errors.

**Comment by ShamsUlAzeem** [comment]
@zjffdu there are a couple of tests that I want to do with the repo for checking out the desired output. Also, have to resolve some of the review comments here then we'll be good to merge this.

**Comment by ShamsUlAzeem** [comment]
@zjffdu hmm seems like the integration tests are failing because I upgraded the guava version to 22.0 from 20.0. They have remove the toStringHelper from Objects and have moved it to MoreObjects starting from version 21.0. I'll go and downgrade the version in the whole project to see if it can run the tests successfully.

**Comment by zjffdu** [comment]
ping @ShamsUlAzeem Any update ?

**Comment by zjffdu** [comment]
Thanks for the hard work and keep me updated. @ShamsUlAzeem

**Comment by zjffdu** [comment]
@ShamsUlAzeem Are you still working on this ? If not, I will take over it and will make it available in 0.9

**Comment by ShamsUlAzeem** [comment]
@zjffdu sure you can take it over from here. It's kinda difficult for me to find the time for this PR

**Comment by zjffdu** [comment]
Thanks @ShamsUlAzeem  really appreciated your contribution
